src/virtualship/utils.py

```diff
@@ -37,3 +40,99 @@ def _dump_yaml(model: BaseModel, stream: TextIO) -> str | None:
 def _generic_load_yaml(data: str, model: BaseModel) -> BaseModel:
     """Load a yaml string into a pydantic model."""
     return model.model_validate(yaml.safe_load(data))
+
+
+def mfp_to_yaml(excel_file_path: str, save_directory: str):
+    """
+    Generates a YAML file (`schedule.yaml`) with spatial and temporal information based on instrument data from MFP excel file.
+
+    Parameters
+    ----------
+    - excel_file_path (str): Path to the Excel file containing coordinate and instrument data.
+    - save_directory (str): Directory where `schedule.yaml` will be saved.
+
+    The function:
+    1. Reads instrument and location data from the Excel file.
+    2. Determines the maximum depth and buffer based on the instruments present.
+    3. Ensures longitude and latitude values remain valid after applying buffer adjustments.
+    4. Saves `schedule.yaml` in the specified directory.
+
+    """
+    # Read data from Excel
+    coordinates_data = pd.read_excel(
+        excel_file_path,
+        usecols=["Station Type", "Name", "Latitude", "Longitude", "Instrument"],
+    )
+    coordinates_data = coordinates_data.dropna()
+
+    # Define maximum depth and buffer for each instrument
+    instrument_properties = {
+        "CTD": {"depth": 5000, "buffer": 1},
+        "DRIFTER": {"depth": 1, "buffer": 5},
+        "ARGO_FLOAT": {"depth": 2000, "buffer": 5},
+    }
+
+    # Extract unique instruments from dataset
+    unique_instruments = np.unique(
+        np.hstack(coordinates_data["Instrument"].apply(lambda a: a.split(", ")).values)
+    )
```

Comment:
I think this would just be clearer as a for loop.

```py
unique_instruments = set()
for row in ....:
    instruments = ...split(", ")
    unique_instruments |= set(instruments)  # or just `unique_instruments = unique_instruments | set(instruments)`
```
That way there's no need to write the comment since its evident from the code